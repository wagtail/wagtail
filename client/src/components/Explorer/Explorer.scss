$c-explorer-bg: #4c4e4d;
$c-explorer-bg-dark: $color-grey-1;
$c-explorer-bg-active: rgba(0, 0, 0, 0.425);
$c-explorer-secondary: #a5a5a5;
$c-explorer-easing: cubic-bezier(0.075, 0.82, 0.165, 1);
$menu-footer-height: 50px;

@import 'ExplorerItem';

.explorer__wrapper,
.explorer__wrapper * {
    box-sizing: border-box;
}

.explorer__wrapper {
    display: flex;
    flex: 1;
}

.explorer {
    width: 100%;
    display: flex;
    flex-direction: column;

    @include medium {
        width: 485px;
        height: 100vh;
        position: fixed;
        z-index: 500;
        left: $menu-width;
    }
}

.c-explorer {
    flex: 1;
    position: relative;
    overflow: hidden;
    background: $c-explorer-bg;

    @include medium {
        box-shadow: 2px 2px 5px $c-explorer-bg-active;
    }
}

.c-explorer > .c-transition-group {
    display: flex;
    flex-direction: column;
    height: 100%;
    z-index: 150;
}

.c-explorer__close {
    padding: 1em;
    color: $c-explorer-secondary;
    border-bottom: 1px solid rgba(200, 200, 200, 0.1);
    cursor: pointer;

    &:focus {
        background-color: $c-explorer-bg-active;
        color: $color-white;
        outline: none;
    }

    // Overrides for default link hover.
    &:hover {
        color: $c-explorer-secondary;
    }

    @include small {
        .explorer-open & {
            display: block;
        }
    }
}

.c-explorer__drawer {
    flex: 1;
    overflow-y: auto;
    -webkit-overflow-scrolling: touch;
}

.c-explorer__header {
    display: block;
    background-color: $c-explorer-bg-dark;
    border-bottom: 1px solid $c-explorer-bg-dark;
    color: $color-white;

    &:focus {
        background-color: $c-explorer-bg-active;
        color: $color-white;
        outline: none;
    }

    // Overrides for default link hover.
    &:hover {
        color: $color-white;
    }

    @include hover {
        background-color: $c-explorer-bg-active;
    }
}

.c-explorer__header__inner {
    padding: 1em 0.75em;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;

    .icon {
        color: $c-explorer-secondary;
        margin-right: 0.25rem;
        font-size: 1rem;
    }

    @include medium {
        padding: 1em 1.5em;
    }
}

.c-explorer__placeholder {
    padding: 1em;
    color: $color-white;

    @include medium {
        padding: 1em 1.75em;
    }
}

.c-explorer__see-more {
    display: block;
    padding: 1em;
    background: rgba(0, 0, 0, 0.3);
    color: $color-white;

    &:focus {
        color: $c-explorer-secondary;
        background: $c-explorer-bg-active;
        outline: none;
    }

    // Overrides for default link hover.
    &:hover {
        color: $color-white;
    }

    @include hover {
        background: $c-explorer-bg-active;
    }

    @include medium {
        padding: 1em 1.75em;
        height: $menu-footer-height;
    }
}
