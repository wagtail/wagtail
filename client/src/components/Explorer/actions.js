import * as admin from '../../api/admin';
import { createAction } from '../../utils/actions';
import { MAX_EXPLORER_PAGES } from '../../config/wagtailConfig';

const getPageStart = createAction('GET_PAGE_START');
const getPageSuccess = createAction('GET_PAGE_SUCCESS', (id, data) => ({ id, data }));
const getPageFailure = createAction('GET_PAGE_FAILURE', (id, error) => ({ id, error }));

/**
 * Gets a page from the API.
 */
function getPage(id) {
  return (dispatch) => {
    dispatch(getPageStart(id));

    return admin.getPage(id).then((data) => {
      dispatch(getPageSuccess(id, data));
    }, (error) => {
      dispatch(getPageFailure(id, error));
    });
  };
}

const getChildrenStart = createAction('GET_CHILDREN_START', id => ({ id }));
const getChildrenSuccess = createAction('GET_CHILDREN_SUCCESS', (id, items, meta) => ({ id, items, meta }));
const getChildrenFailure = createAction('GET_CHILDREN_FAILURE', (id, error) => ({ id, error }));

/**
 * Gets the children of a node from the API.
 */
function getChildren(id, offset = 0) {
  return (dispatch) => {
    dispatch(getChildrenStart(id));

    return admin.getPageChildren(id, {
      offset: offset,
    }).then(({ items, meta }) => {
      const nbPages = offset + items.length;
      dispatch(getChildrenSuccess(id, items, meta));

      // Load more pages if necessary. Only one request is created even though
      // more might be needed, thus naturally throttling the loading.
      if (nbPages < meta.total_count && nbPages < MAX_EXPLORER_PAGES) {
        dispatch(getChildren(id, nbPages));
      }
    }, (error) => {
      dispatch(getChildrenFailure(id, error));
    });
  };
}

const openExplorer = createAction('OPEN_EXPLORER', id => ({ id }));
export const closeExplorer = createAction('CLOSE_EXPLORER');

export function toggleExplorer(id) {
  return (dispatch, getState) => {
    const { explorer, nodes } = getState();

    if (explorer.isVisible) {
      dispatch(closeExplorer());
    } else {
      const page = nodes[id];

      dispatch(openExplorer(id));

      if (!page) {
        dispatch(getChildren(id));
      }

      // We need to get the title of the starting page, only if it is not the site's root.
      const isNotRoot = id !== 1;
      if (isNotRoot) {
        dispatch(getPage(id));
      }
    }
  };
}

export const popPage = createAction('POP_PAGE');
const pushPagePrivate = createAction('PUSH_PAGE', id => ({ id }));

export function pushPage(id) {
  return (dispatch, getState) => {
    const { nodes } = getState();
    const page = nodes[id];

    dispatch(pushPagePrivate(id));

    if (page && !page.isFetching && !page.children.count > 0) {
      dispatch(getChildren(id));
    }
  };
}
