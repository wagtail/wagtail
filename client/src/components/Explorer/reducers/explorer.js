const defaultState = {
  isVisible: false,
  path: [],
};

/**
 * Oversees the state of the explorer. Defines:
 * - Where in the page tree the explorer is at.
 * - Whether the explorer is open or not.
 */
export default function explorer(prevState = defaultState, { type, payload }) {
  switch (type) {
  case 'OPEN_EXPLORER':
    // Provide a starting page when opening the explorer.
    return {
      isVisible: true,
      path: [payload.id],
    };

  case 'CLOSE_EXPLORER':
    return defaultState;

  case 'PUSH_PAGE':
    return {
      isVisible: prevState.isVisible,
      path: prevState.path.concat([payload.id]),
    };

  case 'POP_PAGE':
    return {
      isVisible: prevState.isVisible,
      path: prevState.path.slice(0, -1),
    };

  default:
    return prevState;
  }
}
