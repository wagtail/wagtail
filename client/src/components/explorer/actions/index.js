import { createAction } from 'redux-actions';

import { API_PAGES, PAGES_ROOT_ID } from 'config';

function _getHeaders() {
  const headers = new Headers();
  headers.append('Content-Type', 'application/json');

  return {
    credentials: 'same-origin',
    headers: headers,
    method: 'GET'
  };
}

function _get(url) {
  return fetch(url, _getHeaders()).then(response => response.json());
}

export const fetchStart = createAction('FETCH_START');

export const fetchSuccess = createAction('FETCH_SUCCESS', (id, body) => {
  return { id, body };
});

export const fetchFailure = createAction('FETCH_FAILURE');

export const pushPage = createAction('PUSH_PAGE');

export const popPage = createAction('POP_PAGE');

export const fetchBranchSuccess = createAction('FETCH_BRANCH_SUCCESS', (id, json) => {
  return { id, json };
});

export const fetchBranchStart = createAction('FETCH_BRANCH_START');

export const clearError = createAction('CLEAR_TRANSPORT_ERROR');

export const resetTree = createAction('RESET_TREE');

export const treeResolved = createAction('TREE_RESOLVED');

// Make this a bit better... hmm....
export function fetchTree(id = 1) {
  return (dispatch) => {
    dispatch(fetchBranchStart(id));

    return _get(`${API_PAGES}${id}/`)
      .then(json => {
        dispatch(fetchBranchSuccess(id, json));

        // Recursively walk up the tree to the root, to figure out how deep
        // in the tree we are.
        if (json.meta.parent) {
          dispatch(fetchTree(json.meta.parent.id));
        } else {
          dispatch(treeResolved());
        }
      });
  };
}

export function fetchRoot() {
  return (dispatch) => {
    // TODO Should not need an id.
    dispatch(resetTree(1));

    return _get(`${API_PAGES}?child_of=${PAGES_ROOT_ID}`)
      .then(json => {
        // TODO right now, only works for a single homepage.
        // TODO What do we do if there is no homepage?
        const rootId = json.items[0].id;

        dispatch(fetchTree(rootId));
      });
  };
}

export const toggleExplorer = createAction('TOGGLE_EXPLORER');

export const fetchChildrenSuccess = createAction('FETCH_CHILDREN_SUCCESS', (id, json) => {
  return { id, json };
});

export const fetchChildrenStart = createAction('FETCH_CHILDREN_START');

/**
 * Gets the children of a node from the API
 */
export function fetchChildren(id = 'root') {
  return (dispatch, getState) => {
    const { explorer } = getState();

    let api = `${API_PAGES}?child_of=${id}`;

    if (explorer.fields) {
      api += `&fields=${explorer.fields.map(global.encodeURIComponent).join(',')}`;
    }

    if (explorer.filter) {
      api = `${api}&${explorer.filter}`;
    }

    dispatch(fetchChildrenStart(id));

    return _get(api)
      .then(json => dispatch(fetchChildrenSuccess(id, json)));
  };
}

export function setFilter(filter) {
  return (dispatch, getState) => {
    const { explorer } = getState();
    const id = explorer.path[explorer.path.length - 1];

    dispatch({
      payload: {
        filter,
        id
      },
      type: 'SET_FILTER'
    });

    dispatch(fetchChildren(id));
  };
}

/**
 * TODO: determine if page is already loaded, don't load it again, just push.
 */
export function fetchPage(id = 1) {
  return dispatch => {
    dispatch(fetchStart(id));
    return _get(`${API_PAGES}${id}/`)
      .then(json => dispatch(fetchSuccess(id, json)))
      .then(json => dispatch(fetchChildren(id, json)))
      .catch(json => dispatch(fetchFailure(new Error(JSON.stringify(json)))));
  };
}

export const setDefaultPage = createAction('SET_DEFAULT_PAGE');
