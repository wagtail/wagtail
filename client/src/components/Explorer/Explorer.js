import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';

import * as actions from './actions';

import ExplorerPanel from './ExplorerPanel';

const Explorer = ({
  isVisible,
  nodes,
  path,
  pushPage,
  popPage,
  onClose,
}) => {
  const page = nodes[path[path.length - 1]];

  return isVisible ? (
    <ExplorerPanel
      path={path}
      page={page}
      nodes={nodes}
      onClose={onClose}
      popPage={popPage}
      pushPage={pushPage}
    />
  ) : null;
};

Explorer.propTypes = {
  isVisible: PropTypes.bool.isRequired,
  path: PropTypes.array.isRequired,
  nodes: PropTypes.object.isRequired,

  pushPage: PropTypes.func.isRequired,
  popPage: PropTypes.func.isRequired,
  onClose: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  isVisible: state.explorer.isVisible,
  path: state.explorer.path,
  nodes: state.nodes,
});

const mapDispatchToProps = (dispatch) => ({
  pushPage: (id) => dispatch(actions.pushPage(id)),
  popPage: () => dispatch(actions.popPage()),
  onClose: () => dispatch(actions.closeExplorer()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Explorer);
