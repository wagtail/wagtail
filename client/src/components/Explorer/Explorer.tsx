/* eslint-disable react/prop-types */

import React from 'react';
import { connect } from 'react-redux';

import * as actions from './actions';
import { State as NodeState } from './reducers/nodes';

import ExplorerPanel from './ExplorerPanel';

interface ExplorerProps {
  isVisible: boolean;
  path: number[],
  nodes: NodeState,
  onClose(): void;
  popPage(): void;
  pushPage(id: number): void;
}

const Explorer: React.FunctionComponent<ExplorerProps> = ({
  isVisible,
  nodes,
  path,
  pushPage,
  popPage,
  onClose,
}) => {
  const page = nodes[path[path.length - 1]];

  return isVisible ? (
    <ExplorerPanel
      path={path}
      page={page}
      nodes={nodes}
      onClose={onClose}
      popPage={popPage}
      pushPage={pushPage}
    />
  ) : null;
};

const mapStateToProps = (state) => ({
  isVisible: state.explorer.isVisible,
  path: state.explorer.path,
  nodes: state.nodes,
});

const mapDispatchToProps = (dispatch) => ({
  pushPage: (id) => dispatch(actions.pushPage(id)),
  popPage: () => dispatch(actions.popPage()),
  onClose: () => dispatch(actions.closeExplorer()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Explorer);
