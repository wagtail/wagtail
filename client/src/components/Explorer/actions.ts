import { ThunkAction } from 'redux-thunk';

import * as admin from '../../api/admin';
import { createAction } from '../../utils/actions';
import { MAX_EXPLORER_PAGES } from '../../config/wagtailConfig';

import { State as ExplorerState, Action as ExplorerAction } from './reducers/explorer';
import { State as NodeState, WagtailPageAPI, Action as NodeAction } from './reducers/nodes';

interface State {
  explorer: ExplorerState,
  nodes: NodeState,
}

type Action = ExplorerAction | NodeAction;
type ThunkActionType = ThunkAction<void, State, unknown, Action>;

const getPageSuccess = createAction('GET_PAGE_SUCCESS', (id: number, data: WagtailPageAPI) => ({ id, data }));
const getPageFailure = createAction('GET_PAGE_FAILURE', (id: number, error: Error) => ({ id, error }));

/**
 * Gets a page from the API.
 */
function getPage(id: number): ThunkActionType {
  return (dispatch) => admin.getPage(id).then((data) => {
    dispatch(getPageSuccess(id, data));
  }, (error) => {
    dispatch(getPageFailure(id, error));
  });
}

const getChildrenStart = createAction('GET_CHILDREN_START', (id: number) => ({ id }));
const getChildrenSuccess = createAction(
  'GET_CHILDREN_SUCCESS',
  (id, items: WagtailPageAPI[], meta: any) => ({ id, items, meta })
);
const getChildrenFailure = createAction('GET_CHILDREN_FAILURE', (id: number, error: Error) => ({ id, error }));

/**
 * Gets the children of a node from the API.
 */
function getChildren(id: number, offset = 0): ThunkActionType {
  return (dispatch) => {
    dispatch(getChildrenStart(id));

    return admin.getPageChildren(id, {
      offset: offset,
    }).then(({ items, meta }) => {
      const nbPages = offset + items.length;
      dispatch(getChildrenSuccess(id, items, meta));

      // Load more pages if necessary. Only one request is created even though
      // more might be needed, thus naturally throttling the loading.
      if (nbPages < meta.total_count && nbPages < MAX_EXPLORER_PAGES) {
        dispatch(getChildren(id, nbPages));
      }
    }, (error) => {
      dispatch(getChildrenFailure(id, error));
    });
  };
}

const openExplorer = createAction('OPEN_EXPLORER', id => ({ id }));
export const closeExplorer = createAction('CLOSE_EXPLORER');

export function toggleExplorer(id: number): ThunkActionType {
  return (dispatch, getState) => {
    const { explorer, nodes } = getState();

    if (explorer.isVisible) {
      dispatch(closeExplorer());
    } else {
      const page = nodes[id];

      dispatch(openExplorer(id));

      if (!page) {
        dispatch(getChildren(id));
      }

      // We need to get the title of the starting page, only if it is not the site's root.
      const isNotRoot = id !== 1;
      if (isNotRoot) {
        dispatch(getPage(id));
      }
    }
  };
}

export const popPage = createAction('POP_PAGE');
const pushPagePrivate = createAction('PUSH_PAGE', (id: number) => ({ id }));

export function pushPage(id: number): ThunkActionType {
  return (dispatch, getState) => {
    const { nodes } = getState();
    const page = nodes[id];

    dispatch(pushPagePrivate(id));

    if (page && !page.isFetching && !(page.children.count > 0)) {
      dispatch(getChildren(id));
    }
  };
}
