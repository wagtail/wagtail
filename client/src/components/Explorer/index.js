import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { createStore, combineReducers, applyMiddleware, compose } from 'redux';
import thunkMiddleware from 'redux-thunk';

// import { perfMiddleware } from '../../utils/performance';
import Explorer from './Explorer';
import ExplorerToggle from './ExplorerToggle';
import explorer from './reducers/explorer';
import nodes from './reducers/nodes';

/**
 * Initialises the explorer component on the given nodes.
 */
const initExplorer = (explorerNode, toggleNode) => {
  const rootReducer = combineReducers({
    explorer,
    nodes,
  });

  const middleware = [
    thunkMiddleware,
  ];

  // Uncomment this to use performance measurements.
  // if (process.env.NODE_ENV !== 'production') {
  //   middleware.push(perfMiddleware);
  // }

  const store = createStore(rootReducer, {}, compose(
    applyMiddleware(...middleware),
    // Expose store to Redux DevTools extension.
    window.devToolsExtension ? window.devToolsExtension() : func => func
  ));

  const startPage = parseInt(toggleNode.getAttribute('data-explorer-start-page'), 10);

  ReactDOM.render((
    <Provider store={store}>
      <ExplorerToggle startPage={startPage}>{toggleNode.textContent}</ExplorerToggle>
    </Provider>
  ), toggleNode.parentNode);

  ReactDOM.render((
    <Provider store={store}>
      <Explorer />
    </Provider>
  ), explorerNode);
};

export default Explorer;

export {
  ExplorerToggle,
  initExplorer,
};
