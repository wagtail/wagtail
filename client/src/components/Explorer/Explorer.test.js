import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware, combineReducers } from 'redux';
import thunkMiddleware from 'redux-thunk';
import * as actions from './actions';
import explorer from './reducers/explorer';
import nodes from './reducers/nodes';
import Explorer from './Explorer';

const rootReducer = combineReducers({
  explorer,
  nodes,
});

const store = createStore(rootReducer, {}, applyMiddleware(thunkMiddleware));

describe('Explorer', () => {
  it('exists', () => {
    expect(Explorer).toBeDefined();
  });

  it('renders', () => {
    expect(shallow(<Explorer store={store} />)).toMatchSnapshot();
    expect(shallow(<Provider store={store}><Explorer /></Provider>)).toMatchSnapshot();
  });

  it('visible', () => {
    store.dispatch(actions.toggleExplorer(1));
    expect(shallow(<Explorer store={store} />)).toMatchSnapshot();
    expect(shallow(<Explorer store={store} />).dive()).toMatchSnapshot();
  });

  describe('actions', () => {
    let wrapper;

    beforeEach(() => {
      store.dispatch = jest.fn();
      wrapper = shallow(<Explorer store={store} />);
    });

    it('pushPage', () => {
      wrapper.prop('pushPage')();
      expect(store.dispatch).toHaveBeenCalled();
    });

    it('popPage', () => {
      wrapper.prop('popPage')();
      expect(store.dispatch).toHaveBeenCalled();
    });

    it('onClose', () => {
      wrapper.prop('onClose')();
      expect(store.dispatch).toHaveBeenCalled();
    });
  });
});
