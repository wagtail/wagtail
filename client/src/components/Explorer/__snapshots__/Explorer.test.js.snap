// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explorer renders 1`] = `
<Explorer
  isVisible={false}
  nodes={Object {}}
  onClose={[Function]}
  path={Array []}
  popPage={[Function]}
  pushPage={[Function]}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;

exports[`Explorer renders 2`] = `<Connect(Explorer) />`;

exports[`Explorer visible 1`] = `
<Explorer
  isVisible={true}
  nodes={
    Object {
      "1": Object {
        "children": Object {
          "count": 0,
          "items": Array [],
        },
        "id": 0,
        "isError": false,
        "isFetching": true,
        "meta": Object {
          "children": Object {},
          "status": Object {
            "has_unpublished_changes": true,
            "live": false,
            "status": "",
          },
        },
      },
    }
  }
  onClose={[Function]}
  path={
    Array [
      1,
    ]
  }
  popPage={[Function]}
  pushPage={[Function]}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;

exports[`Explorer visible 2`] = `
<ExplorerPanel
  nodes={
    Object {
      "1": Object {
        "children": Object {
          "count": 0,
          "items": Array [],
        },
        "id": 0,
        "isError": false,
        "isFetching": true,
        "meta": Object {
          "children": Object {},
          "status": Object {
            "has_unpublished_changes": true,
            "live": false,
            "status": "",
          },
        },
      },
    }
  }
  onClose={[Function]}
  page={
    Object {
      "children": Object {
        "count": 0,
        "items": Array [],
      },
      "id": 0,
      "isError": false,
      "isFetching": true,
      "meta": Object {
        "children": Object {},
        "status": Object {
          "has_unpublished_changes": true,
          "live": false,
          "status": "",
        },
      },
    }
  }
  path={
    Array [
      1,
    ]
  }
  popPage={[Function]}
  pushPage={[Function]}
/>
`;
